# Lab 5: PCA Using Eigenvectors and Singular Value Decomposition

# First: you should run this cell

In [1]:
# Make all numpy available via shorter 'np' prefix
import numpy as np

# Question 1 : PCA Using Eigenvectors

#### Create a matrix contains the following data

| | Feature #1 | Feature #2 | Feature #3 | Feature #4 |
| --- | --- | --- | --- | --- |
| **Sample #1** | 1 | 2 | 3 | 4 |
| **Sample #2** | 5 | 5 | 6 | 7 |
| **Sample #3**| 1 | 4 | 2 | 3 |
| **Sample #4**| 5 | 3 | 2 | 1 |
| **Sample #5**| 8 | 1 | 2 | 2 |

In [19]:
M = np.array([[2, 1],
              [3, 5],
              [4, 3],
              [5, 6],
              [6, 7],
              [7, 8]])

Mean = np.mean(M, axis=0)
print("\nData mean = \n", Mean)
Std = np.std(M, axis=0, ddof=1)
print("\nData standard deviation = \n", Std)
X = (M - Mean) / Std
print("\nX = \n", X)
C = (X.transpose() @ X)/4
print("C = \n", C)
w, V = np.linalg.eig(C)
print("\nEigenvalues = \n", w)
print("\nEigenvectors = \n",V)
eig_vecs_2 = V[:, :1]
print("\neig_vecs_2 = \n", eig_vecs_2)
M_transformed = X @ eig_vecs_2
print("\nM_transformed = \n", M_transformed)


Data mean = 
 [4.5 5. ]

Data standard deviation = 
 [1.87082869 2.60768096]

X = 
 [[-1.33630621 -1.53392998]
 [-0.80178373  0.        ]
 [-0.26726124 -0.76696499]
 [ 0.26726124  0.38348249]
 [ 0.80178373  0.76696499]
 [ 1.33630621  1.15044748]]
C = 
 [[1.25       1.12739008]
 [1.12739008 1.25      ]]

Eigenvalues = 
 [2.37739008 0.12260992]

Eigenvectors = 
 [[ 0.70710678 -0.70710678]
 [ 0.70710678  0.70710678]]

eig_vecs_2 = 
 [[0.70710678]
 [0.70710678]]

M_transformed = 
 [[-2.02956347]
 [-0.56694671]
 [-0.73130838]
 [ 0.46014531]
 [ 1.10927285]
 [ 1.7584004 ]]


In [11]:
M = np.array([[1,2,3,4],
              [5,5,6,7],
              [1,4,2,3],
              [5,3,2,1],
              [8,1,2,2]])
print('M = \n{}\n'.format(M))

M = 
[[1 2 3 4]
 [5 5 6 7]
 [1 4 2 3]
 [5 3 2 1]
 [8 1 2 2]]

M = 
 [[1 2 3 4]
 [5 5 6 7]
 [1 4 2 3]
 [5 3 2 1]
 [8 1 2 2]]

Data mean = 
 [4.  3.  3.  3.4]

Data standard deviation = 
 [3.         1.58113883 1.73205081 2.30217289]

X = 
 [[-1.         -0.63245553  0.          0.26062335]
 [ 0.33333333  1.26491106  1.73205081  1.56374007]
 [-1.          0.63245553 -0.57735027 -0.1737489 ]
 [ 0.33333333  0.         -0.57735027 -1.04249338]
 [ 1.33333333 -1.26491106 -0.57735027 -0.60812114]]
C = 
 [[ 1.         -0.31622777  0.04811252 -0.18098843]
 [-0.31622777  1.          0.63900965  0.61812254]
 [ 0.04811252  0.63900965  1.          0.94044349]
 [-0.18098843  0.61812254  0.94044349  1.        ]]


### Step 1: Standardize the dataset.
- Calculate data mean for each feature.
- Calculate data standard deviation for each feature.

| | Feature #1 | Feature #2 | Feature #3 | Feature #4 |
| --- | --- | --- | --- | --- |
| **$\mu$** | 4 | 3 | 3 | 3.4 |
| **$\sigma$**| 3 | 1.58114 | 1.73205 | 2.30217 |


- Standardize the data.

| | Feature #1 | Feature #2 | Feature #3 | Feature #4 |
| --- | --- | --- | --- | --- |
| **Sample #1** | -1.         | -0.63245553 |  0.         |  0.26062335 |
| **Sample #2** |  0.33333333 | 1.26491106 |  1.73205081 |  1.56374007 |
| **Sample #3** | -1.         | 0.63245553 | -0.57735027 | -0.1737489  |
| **Sample #4** |  0.33333333 | 0.         | -0.57735027 | -1.04249338 |
| **Sample #5** |  1.33333333 | -1.26491106 | -0.57735027 | -0.60812114 |




In [73]:
#type your code here:
print("M = \n", M)
Mean = np.mean(M, axis=0)
print("\nData mean = \n", Mean)
Std = np.std(M, axis=0, ddof=1)
print("\nData standard deviation = \n", Std)
X = (M - Mean) / Std
print("\nX = \n", X)

M = 
 [[1 2 3 4]
 [5 5 6 7]
 [1 4 2 3]
 [5 3 2 1]
 [8 1 2 2]]

Data mean = 
 [4.  3.  3.  3.4]

Data standard deviation = 
 [3.         1.58113883 1.73205081 2.30217289]

X = 
 [[-1.         -0.63245553  0.          0.26062335]
 [ 0.33333333  1.26491106  1.73205081  1.56374007]
 [-1.          0.63245553 -0.57735027 -0.1737489 ]
 [ 0.33333333  0.         -0.57735027 -1.04249338]
 [ 1.33333333 -1.26491106 -0.57735027 -0.60812114]]


### Step 2: Calculate the covariance matrix for the features in the dataset.

-  Calculate covariance matrix using the formula: $C = \frac{X^TX}{n-1}$,  where $n$ is the number of data points.
##### Note: This formula is used because the data has <b>zero</b> mean and <b>1</b> standard deviation.

| | Feature #1 | Feature #2 | Feature #3 | Feature #4 |
| --- | --- | --- | --- | --- |
|**Feature #1** |  1.         | -0.31622777 |  0.04811252 | -0.18098843 |
|**Feature #2** | -0.31622777 |  1.         |  0.63900965 |  0.61812254 |
|**Feature #3** |  0.04811252 |  0.63900965 |  1.         |  0.94044349 |
|**Feature #4** | -0.18098843 |  0.61812254 |  0.94044349 |  1.         |

In [74]:
#type your code here:
C = (X.transpose() @ X)/4
print("C = \n", C)

C = 
 [[ 1.         -0.31622777  0.04811252 -0.18098843]
 [-0.31622777  1.          0.63900965  0.61812254]
 [ 0.04811252  0.63900965  1.          0.94044349]
 [-0.18098843  0.61812254  0.94044349  1.        ]]


### Step 3: Calculate the eigenvalues and eigenvectors for the covariance matrix and sort eigenvalues and their corresponding eigenvectors.

- Find eigen values and the corresponding eigen vectors for the covariance matrix
- Calculate how much variance captured in each component. $\frac{\lambda_j}{\sum_{k=1}^{m}\lambda_k}$, where $m$ is the number of features.

| Eigenvector #1 | Eigenvector #2 | Eigenvector #3 | Eigenvector #4 |
| --- | --- | --- | --- |
| 0.16195986 | -0.91705888 | -0.30707099 |  0.19616173 |
|-0.52404813 |  0.20692161 | -0.81731886 |  0.12061043 |
|-0.58589647 | -0.3205394  |  0.1882497  | -0.72009851 |
|-0.59654663 | -0.11593512 |  0.44973251 |  0.65454704 |

In [75]:
#type your code here:
w, V = np.linalg.eig(C)
print("Eigenvalues = \n", w)
var = []
for i in w:
     var.append((i/sum(w))) 
print("\npercentage of variance = \n", var)
print("\nEigenvectors = \n",V)

Eigenvalues = 
 [2.51579324 1.0652885  0.39388704 0.02503121]

percentage of variance = 
 [0.6289483102024862, 0.26632212587703585, 0.09847176095061148, 0.0062578029698664266]

Eigenvectors = 
 [[ 0.16195986 -0.91705888 -0.30707099  0.19616173]
 [-0.52404813  0.20692161 -0.81731886  0.12061043]
 [-0.58589647 -0.3205394   0.1882497  -0.72009851]
 [-0.59654663 -0.11593512  0.44973251  0.65454704]]


### Step 4: Pick $k$ eigenvalues and form a matrix of eigenvectors.

- Select the first two eigen vectors

In [76]:
#type your code here:
eig_vecs_2 = V[:, :2]
print("eig_vecs_2 = \n", eig_vecs_2)

eig_vecs_2 = 
 [[ 0.16195986 -0.91705888]
 [-0.52404813  0.20692161]
 [-0.58589647 -0.3205394 ]
 [-0.59654663 -0.11593512]]


### Step 5:Transform the standardized matrix matrix.

- `M_transformed = X @ eig_vecs_2`

In [77]:
#type your code here:
M_transformed = X @ eig_vecs_2
print("M_transformed = \n", M_transformed)

M_transformed = 
 [[ 1.40033078e-02  7.55974765e-01]
 [-2.55653399e+00 -7.80431775e-01]
 [-5.14801919e-02  1.25313470e+00]
 [ 1.01415002e+00  2.38808310e-04]
 [ 1.57986086e+00 -1.22891650e+00]]


# Question 2 : Singular Value Decomposition (SVD)

#### For a square (real) matrices of size $(n,n)$ can be "eigendecomposed"  :
$$A = V . Diag(\Lambda). V^T$$

With $V$: eigenvectors 
; $Diag(\Lambda)$: eigenvalues (on the diagonal).

#### But for a non square  of size $(n,m) (n \neq m)$, we need to use the  Singular Value Decomposition (SVD)

- Let $A $ be a matrix of size $(n,m)$ with  $n \neq m$ the SVD of A is defined by:

$$A= U . S . V^T $$

Where:
 $U$: size $(n,n)$, square;
 $V$: size $(m,m)$, square;
 $S$: size $(n,m)$, rectanguler if $n \neq m$
 
- The values on the "diagonal" of $S$ are called: **singular values**. The singular values are generally sorted by decreasing order.
- The columns of $U$ are called the **left-singular vectors**
- The columns of $V$ are called the **right-singular vectors**

#### (a) For the matrix $A$, find the SVD

In [78]:
A = np.array([[1, 1, 1],
              [3, 4, 5],
              [4, 4, 4],                            
              [4, 5, 4]                      
              ])
rank = np.linalg.matrix_rank(A)
n, m = A.shape
print("n = ", n)
print("---------------------------------------------")
print("m = ", m)
print("---------------------------------------------")
print("rank = ", rank)
print("---------------------------------------------")
print("A = \n", A)

n =  4
---------------------------------------------
m =  3
---------------------------------------------
rank =  3
---------------------------------------------
A = 
 [[1 1 1]
 [3 4 5]
 [4 4 4]
 [4 5 4]]


- `U, s, VT = np.linalg.svd(A, full_matrices=True)`

In [79]:
#type your code here:
U, s, VT = np.linalg.svd(A, full_matrices=True)
print("U = \n", U)
print("\ns = \n", s)
print("\nVT = \n", VT)

U = 
 [[-1.38201662e-01 -6.54749524e-02 -1.88246808e-01 -9.70142500e-01]
 [-5.59976765e-01  8.18823566e-01  1.26309104e-01 -1.33226763e-15]
 [-5.52806648e-01 -2.61899810e-01 -7.52987231e-01  2.42535625e-01]
 [-6.01440881e-01 -5.06605850e-01  6.17753494e-01 -1.11022302e-15]]

s = 
 [12.49789499  1.23098537  0.53599988]

VT = 
 [[-0.51489647 -0.60782554 -0.60450785]
 [-0.55486191 -0.30122956  0.77549276]
 [-0.65345993  0.73471687 -0.1821572 ]]


- Construct the matrix $A$ from the SVD:
$$A= U . S . V^T $$
Where:
 $U$: size $(n,n)$, square;
 $V$: size $(m,m)$, square;
 $S$: size $(n,m)$, rectanguler if $n \neq m$

In [80]:
#type your code here:
print("U = \n", U)
S = np.zeros((A.shape[0], A.shape[1]))
S[:A.shape[1], :A.shape[1]] = np.diag(s)
print("\nS = \n", S)
print("\nVT = \n", VT)
print("\nA = \n", A)
print("\nU @ S @ VT = \n", U @ S @ VT)

U = 
 [[-1.38201662e-01 -6.54749524e-02 -1.88246808e-01 -9.70142500e-01]
 [-5.59976765e-01  8.18823566e-01  1.26309104e-01 -1.33226763e-15]
 [-5.52806648e-01 -2.61899810e-01 -7.52987231e-01  2.42535625e-01]
 [-6.01440881e-01 -5.06605850e-01  6.17753494e-01 -1.11022302e-15]]

S = 
 [[12.49789499  0.          0.        ]
 [ 0.          1.23098537  0.        ]
 [ 0.          0.          0.53599988]
 [ 0.          0.          0.        ]]

VT = 
 [[-0.51489647 -0.60782554 -0.60450785]
 [-0.55486191 -0.30122956  0.77549276]
 [-0.65345993  0.73471687 -0.1821572 ]]

A = 
 [[1 1 1]
 [3 4 5]
 [4 4 4]
 [4 5 4]]

U @ S @ VT = 
 [[1. 1. 1.]
 [3. 4. 5.]
 [4. 4. 4.]
 [4. 5. 4.]]


#### $k$-rank approximation

- Construct the matrix $\tilde{A}$ using **$k$-rank approximation**.
- Calculate the approximation error using frobenius norm.

$$\tilde{A}= \tilde{U} . \tilde{S} . \tilde{V}^T $$
Where:
 $\tilde{U}$: size $(n,k)$,
  $\tilde{S}$: size $(k,k)$,
 $\tilde{V}$: size $(k,m)$;


In [82]:
#type your code here:
k = 3
print("k = ", k)
print("\nA = \n", A)
U_tilda, s_tilda, VT_tilda = np.linalg.svd(A, full_matrices=False) 
A_tilda = np.matrix(U_tilda[:, :k]) * np.diag(s_tilda[:k]) * np.matrix(VT_tilda[:k,:])
print("\nA_tilda = \n", A_tilda)
print("\nA_tilde_rank = ", np.linalg.matrix_rank(A_tilda))
print("\nerror = ",np.linalg.norm(A - A_tilda))

k =  3

A = 
 [[1 1 1]
 [3 4 5]
 [4 4 4]
 [4 5 4]]

A_tilda = 
 [[1. 1. 1.]
 [3. 4. 5.]
 [4. 4. 4.]
 [4. 5. 4.]]

A_tilde_rank =  3

error =  7.482277218026854e-15


In [83]:
#type your code here:
k = 2
print("k = ", k)
print("\nA = \n", A) 
A_tilda = np.matrix(U_tilda[:, :k]) * np.diag(s_tilda[:k]) * np.matrix(VT_tilda[:k,:])
print("\nA_tilda = \n", A_tilda)
print("\nA_tilde_rank = ", np.linalg.matrix_rank(A_tilda))
print("\nerror = ",np.linalg.norm(A - A_tilda))

k =  2

A = 
 [[1 1 1]
 [3 4 5]
 [4 4 4]
 [4 5 4]]

A_tilda = 
 [[0.93406572 1.07413313 0.98162029]
 [3.04424033 3.95025845 5.01233235]
 [3.73626287 4.29653251 3.92648116]
 [4.21637091 4.75672364 4.06031513]]

A_tilde_rank =  2

error =  0.5359998808909967


In [84]:
#type your code here:
k = 1
print("k = ", k)
print("\nA = \n", A) 
A_tilda = np.matrix(U_tilda[:, :k]) * np.diag(s_tilda[:k]) * np.matrix(VT_tilda[:k,:])
print("\nA_tilda = \n", A_tilda)
print("\nA_tilde_rank = ", np.linalg.matrix_rank(A_tilda))
print("\nerror = ",np.linalg.norm(A - A_tilda))

k =  1

A = 
 [[1 1 1]
 [3 4 5]
 [4 4 4]
 [4 5 4]]

A_tilda = 
 [[0.88934457 1.04985441 1.044124  ]
 [3.60351884 4.25388574 4.2306668 ]
 [3.55737826 4.19941766 4.17649602]
 [3.87034549 4.56886954 4.54393133]]

A_tilde_rank =  1

error =  1.3426171667971218


#### (b) Repeat the previous steps in (a) for the following matrix

In [85]:
A = np.array([[1, 1, 1, 0, 1],
              [3, 4, 5, 1, 0],
              [4, 4, 4, 0, 1],                            
              [4, 5, 4, 1, 1],                            
              [4, 6, 4, 0, 1],                            
              [4, 7, 4, 1, 1],                            
              [4, 8, 4, 0, 1],                            
              [4, 9, 4, 1, 0],                            
              ])

n, m=A.shape
print("n= ", n)
print("---------------------------------------------")
print("m = ", m)
print("---------------------------------------------")
print("rank = ", rank)
print("---------------------------------------------")
print("A = \n", A)

n=  8
---------------------------------------------
m =  5
---------------------------------------------
rank =  3
---------------------------------------------
A = 
 [[1 1 1 0 1]
 [3 4 5 1 0]
 [4 4 4 0 1]
 [4 5 4 1 1]
 [4 6 4 0 1]
 [4 7 4 1 1]
 [4 8 4 0 1]
 [4 9 4 1 0]]


In [86]:
#type your code here:
U, s, VT = np.linalg.svd(A, full_matrices=True)
print("U = \n", U)
print("\ns = \n", s)
print("\nVT = \n", VT)

U = 
 [[-7.79647196e-02  1.53017197e-01 -3.53029757e-01  2.28776787e-01
  -5.76962582e-01 -3.89374906e-01 -4.47402884e-01  3.29955199e-01]
 [-3.00949266e-01  5.20610942e-01  6.79399402e-01 -2.51753516e-01
  -3.36794308e-01  1.38777878e-16  0.00000000e+00  1.24900090e-16]
 [-3.00576700e-01  4.63617867e-01 -3.27397215e-01 -2.20288109e-01
   4.89461080e-01  1.59353557e-01 -2.77306115e-02  5.22744336e-01]
 [-3.36545919e-01  2.83109965e-01 -4.23712681e-02  5.82223875e-01
   2.17668378e-01 -4.40108902e-01  3.86180307e-01 -2.74945500e-01]
 [-3.66610234e-01  7.50478785e-02 -3.01628285e-01 -3.07975353e-01
   6.53503597e-02  2.40580614e-02 -4.42569805e-01 -6.88054373e-01]
 [-4.02579453e-01 -1.05460024e-01 -1.66023380e-02  4.94536632e-01
  -2.06442342e-01  7.32140081e-01 -5.06281445e-02  2.74791006e-02]
 [-4.32643767e-01 -3.13522110e-01 -2.75859355e-01 -3.95662596e-01
  -3.58760361e-01 -8.60678922e-02  5.82151138e-01  8.28212366e-02]
 [-4.64852261e-01 -5.43513653e-01  3.70740529e-01  2.83843017e-

In [87]:
#type your code here:
print("U = \n", U)
S = np.zeros((A.shape[0], A.shape[1]))
S[:A.shape[1], :A.shape[1]] = np.diag(s)
print("\nS = \n", S)
print("\nVT = \n", VT)
print("\nA = \n", A)
print("\nU @ S @ VT = \n", U @ S @ VT)

U = 
 [[-7.79647196e-02  1.53017197e-01 -3.53029757e-01  2.28776787e-01
  -5.76962582e-01 -3.89374906e-01 -4.47402884e-01  3.29955199e-01]
 [-3.00949266e-01  5.20610942e-01  6.79399402e-01 -2.51753516e-01
  -3.36794308e-01  1.38777878e-16  0.00000000e+00  1.24900090e-16]
 [-3.00576700e-01  4.63617867e-01 -3.27397215e-01 -2.20288109e-01
   4.89461080e-01  1.59353557e-01 -2.77306115e-02  5.22744336e-01]
 [-3.36545919e-01  2.83109965e-01 -4.23712681e-02  5.82223875e-01
   2.17668378e-01 -4.40108902e-01  3.86180307e-01 -2.74945500e-01]
 [-3.66610234e-01  7.50478785e-02 -3.01628285e-01 -3.07975353e-01
   6.53503597e-02  2.40580614e-02 -4.42569805e-01 -6.88054373e-01]
 [-4.02579453e-01 -1.05460024e-01 -1.66023380e-02  4.94536632e-01
  -2.06442342e-01  7.32140081e-01 -5.06281445e-02  2.74791006e-02]
 [-4.32643767e-01 -3.13522110e-01 -2.75859355e-01 -3.95662596e-01
  -3.58760361e-01 -8.60678922e-02  5.82151138e-01  8.28212366e-02]
 [-4.64852261e-01 -5.43513653e-01  3.70740529e-01  2.83843017e-

In [88]:
#type your code here:
k = 5
print("k = ", k)
print("\nA = \n", A)
U_tilda, s_tilda, VT_tilda = np.linalg.svd(A, full_matrices=False) 
A_tilda = np.matrix(U_tilda[:, :k]) * np.diag(s_tilda[:k]) * np.matrix(VT_tilda[:k,:])
print("\nA_tilda = \n", A_tilda)
print("\nA_tilde_rank = ", np.linalg.matrix_rank(A_tilda))
print("\nerror = ",np.linalg.norm(A - A_tilda))

k =  5

A = 
 [[1 1 1 0 1]
 [3 4 5 1 0]
 [4 4 4 0 1]
 [4 5 4 1 1]
 [4 6 4 0 1]
 [4 7 4 1 1]
 [4 8 4 0 1]
 [4 9 4 1 0]]

A_tilda = 
 [[ 1.00000000e+00  1.00000000e+00  1.00000000e+00 -4.40619763e-16
   1.00000000e+00]
 [ 3.00000000e+00  4.00000000e+00  5.00000000e+00  1.00000000e+00
   4.71844785e-16]
 [ 4.00000000e+00  4.00000000e+00  4.00000000e+00 -3.34801631e-16
   1.00000000e+00]
 [ 4.00000000e+00  5.00000000e+00  4.00000000e+00  1.00000000e+00
   1.00000000e+00]
 [ 4.00000000e+00  6.00000000e+00  4.00000000e+00 -4.73145828e-16
   1.00000000e+00]
 [ 4.00000000e+00  7.00000000e+00  4.00000000e+00  1.00000000e+00
   1.00000000e+00]
 [ 4.00000000e+00  8.00000000e+00  4.00000000e+00 -5.81132364e-16
   1.00000000e+00]
 [ 4.00000000e+00  9.00000000e+00  4.00000000e+00  1.00000000e+00
  -1.22124533e-15]]

A_tilde_rank =  5

error =  2.5198508814155332e-14


In [89]:
#type your code here:
k = 4
print("k = ", k)
print("\nA = \n", A) 
A_tilda = np.matrix(U_tilda[:, :k]) * np.diag(s_tilda[:k]) * np.matrix(VT_tilda[:k,:])
print("\nA_tilda = \n", A_tilda)
print("\nA_tilde_rank = ", np.linalg.matrix_rank(A_tilda))
print("\nerror = ",np.linalg.norm(A - A_tilda))

k =  4

A = 
 [[1 1 1 0 1]
 [3 4 5 1 0]
 [4 4 4 0 1]
 [4 5 4 1 1]
 [4 6 4 0 1]
 [4 7 4 1 1]
 [4 8 4 0 1]
 [4 9 4 1 0]]

A_tilda = 
 [[ 1.25922885e+00  9.51492408e-01  8.70593425e-01 -1.36648969e-02
   7.86705318e-01]
 [ 3.15132143e+00  3.97168433e+00  4.92446062e+00  9.92023296e-01
  -1.24507961e-01]
 [ 3.78008551e+00  4.04115099e+00  4.10978092e+00  1.15924939e-02
   1.18094665e+00]
 [ 3.90220176e+00  5.01830027e+00  4.04882070e+00  1.00515530e+00
   1.08046884e+00]
 [ 3.97063813e+00  6.00549427e+00  4.01465739e+00  1.54777096e-03
   1.02415908e+00]
 [ 4.09275439e+00  6.98264355e+00  3.95369718e+00  9.95110578e-01
   9.23681266e-01]
 [ 4.16119076e+00  7.96983756e+00  3.91953386e+00 -8.49695197e-03
   8.67371508e-01]
 [ 3.86436371e+00  9.02538062e+00  4.06770939e+00  1.00714988e+00
   1.11602155e-01]]

A_tilde_rank =  4

error =  0.6296604396728646


In [90]:
#type your code here:
k = 3
print("k = ", k)
print("\nA = \n", A) 
A_tilda = np.matrix(U_tilda[:, :k]) * np.diag(s_tilda[:k]) * np.matrix(VT_tilda[:k,:])
print("\nA_tilda = \n", A_tilda)
print("\nA_tilde_rank = ", np.linalg.matrix_rank(A_tilda))
print("\nerror = ",np.linalg.norm(A - A_tilda))

k =  3

A = 
 [[1 1 1 0 1]
 [3 4 5 1 0]
 [4 4 4 0 1]
 [4 5 4 1 1]
 [4 6 4 0 1]
 [4 7 4 1 1]
 [4 8 4 0 1]
 [4 9 4 1 0]]

A_tilda = 
 [[ 1.21384154  0.96160619  0.94039684 -0.20890102  0.69940153]
 [ 3.20126712  3.96055479  4.84764664  1.20686756 -0.02843599]
 [ 3.82378874  4.03141247  4.04256754  0.19958445  1.26501107]
 [ 3.78669365  5.04403927  4.2264664   0.50829052  0.85828566]
 [ 4.03173774  5.99187927  3.92068925  0.26437126  1.14168593]
 [ 3.99464265  7.00450607  4.10458811  0.57307732  0.73496053]
 [ 4.23968675  7.95234607  3.79881096  0.32915806  1.0183608 ]
 [ 3.85873252  9.02663544  4.07636989  0.98292697  0.10077039]]

A_tilde_rank =  3

error =  1.185236342936831


In [91]:
#type your code here:
k = 2
print("k = ", k)
print("\nA = \n", A) 
A_tilda = np.matrix(U_tilda[:, :k]) * np.diag(s_tilda[:k]) * np.matrix(VT_tilda[:k,:])
print("\nA_tilda = \n", A_tilda)
print("\nA_tilde_rank = ", np.linalg.matrix_rank(A_tilda))
print("\nerror = ",np.linalg.norm(A - A_tilda))

k =  2

A = 
 [[1 1 1 0 1]
 [3 4 5 1 0]
 [4 4 4 0 1]
 [4 5 4 1 1]
 [4 6 4 0 1]
 [4 7 4 1 1]
 [4 8 4 0 1]
 [4 9 4 1 0]]

A_tilda = 
 [[0.97120303 0.97817264 1.17745474 0.14101019 0.2345008 ]
 [3.66822055 3.92867295 4.39143298 0.53346975 0.86625708]
 [3.59876756 4.04677607 4.2624133  0.52408954 0.83386553]
 [3.75757173 5.04602761 4.25491852 0.55028749 0.80248744]
 [3.82442764 6.00603363 4.12323129 0.56333505 0.7444752 ]
 [3.98323181 7.00528516 4.11573651 0.589533   0.71309711]
 [4.05008771 7.96529118 3.98404928 0.60258057 0.65508487]
 [4.1135437  9.00923788 3.82741928 0.61546144 0.58899422]]

A_tilde_rank =  2

error =  2.246527226817545


In [92]:
#type your code here:
k = 1
print("k = ", k)
print("\nA = \n", A) 
A_tilda = np.matrix(U_tilda[:, :k]) * np.diag(s_tilda[:k]) * np.matrix(VT_tilda[:k,:])
print("\nA_tilda = \n", A_tilda)
print("\nA_tilde_rank = ", np.linalg.matrix_rank(A_tilda))
print("\nerror = ",np.linalg.norm(A - A_tilda))

k =  1

A = 
 [[1 1 1 0 1]
 [3 4 5 1 0]
 [4 4 4 0 1]
 [4 5 4 1 1]
 [4 6 4 0 1]
 [4 7 4 1 1]
 [4 8 4 0 1]
 [4 9 4 1 0]]

A_tilda = 
 [[0.79493186 1.31209448 0.84185871 0.1173312  0.14945219]
 [3.06849251 5.06477638 3.24963343 0.45290665 0.57689591]
 [3.06469381 5.05850635 3.24561049 0.45234596 0.57618173]
 [3.43143763 5.66384444 3.63400412 0.50647701 0.64513187]
 [3.73797475 6.16980689 3.95863691 0.5517216  0.70276278]
 [4.10471857 6.77514498 4.34703055 0.60585265 0.77171293]
 [4.41125569 7.28110744 4.67166334 0.65109724 0.82934383]
 [4.73965497 7.823155   5.01944886 0.69956867 0.89108496]]

A_tilde_rank =  1

error =  4.03473633898596
